### PR TITLE
Quic Connection Cache

### PR DESCRIPTION
**Pull request by ryleung-solana**
Add connection cache to allow add modules that send data to get or create connections (e.g. for quic) associated with a certain IP address - right now intended just to get a basic interface established to unblock other quic work; this may need some work to be performant enough and to allow configuration of whether to use quic or UDP.

#### Summary of Changes

Fixes #

### CODE REVIEW

**Review by t-nelson — COMMENTED** [review]
as alluded to in a previous pr, we avoid todos in the code. prefer github issues

File: client/src/connection_cache.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        net::{SocketAddr, UdpSocket},
+        sync::{Arc, Mutex},
+    },
+};
+
+lazy_static! {
+    // TODO: all implementations of TpuConnection should be Sync + Send but make sure...
+    // TODO: figure out a way to get rid of this mutex if possible
+    // try_insert seems promising but it's still nightly-only...
+    static ref CONNECTION_MAP: Mutex<HashMap<SocketAddr, Arc<dyn TpuConnection + 'static + Sync + Send>>> = Mutex::new(HashMap::new());
```
Comment (sakridge) [inline client/src/connection_cache.rs:15]: prefer not to use lazy_static here, but maybe we can change it later if it's significantly easier.

**Comment by ryleung-solana** [comment]
Thanks for the comments @t-nelson @sakridge, I have a first implementation of evicting old connections up but it's pretty bad... any suggestions are welcome.

File: client/src/connection_cache.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    std::{
+        collections::HashMap,
+        net::{SocketAddr, UdpSocket},
+        sync::{Arc, Mutex},
+    },
+};
+
+lazy_static! {
+    // TODO: all implementations of TpuConnection should be Sync + Send but make sure...
+    // TODO: figure out a way to get rid of this mutex if possible
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply client/src/connection_cache.rs]: Seems promising, but with the addition of the BTreeMap and other things for connection eviction, it seems that for now a mutex is needed... getting rid of the mutex if possible would still be ideal, and will keep an eye out for possibilities.

**Comment by sakridge** [comment]
> Thanks for the comments @t-nelson @sakridge, I have a first implementation of evicting old connections up but it's pretty bad... any suggestions are welcome.

ok. seems ok for now. Can we just add some unit test for this logic?

**Comment by ryleung-solana** [comment]
> > Thanks for the comments @t-nelson @sakridge, I have a first implementation of evicting old connections up but it's pretty bad... any suggestions are welcome.
> 
> ok. seems ok for now. Can we just add some unit test for this logic?

Added a test for the caching and connection eviction functionality.

File: client/src/connection_cache.rs (hunk 1 above)
Comment (ryleung-solana) [reply client/src/connection_cache.rs]: Agreed, the use of lazy_static here isn't ideal... will have a followup PR to improve this and other aspects.

**Comment by ryleung-solana** [comment]
> as alluded to in a previous pr, we avoid todos in the code. prefer github issues

Done - all TODOs now refer to Github issues.

File: client/src/connection_cache.rs [hunk 3]
```diff
@@ -2,33 +2,82 @@ use {
...
+struct ConnMap {
+    // Keeps track of the connection associated with an addr and the last time it was used
+    map: HashMap<SocketAddr, (Arc<dyn TpuConnection + 'static + Sync + Send>, u64)>,
+    // Helps to find the least recently used connection. The search and inserts are O(log(n))
+    // but since we're bounding the size of the collections, this should be constant
+    // (and hopefully negligible) time. In theory, we can do this in constant time
+    // with a queue implemented as a doubly-linked list (and all the
+    // HashMap entries holding a "pointer" to the corresponding linked-list node),
+    // so we can push, pop and bump a used connection back to the end of the queue in O(1) time, but
+    // that seems non-"Rust-y" and low bang/buck. This is still pretty terrible though...
```
Comment (t-nelson) [inline client/src/connection_cache.rs:23]: Did you consider [`lru`](https://docs.rs/lru/latest/lru/index.html)?

File: client/src/connection_cache.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    map.ticks += 1;
+    conn
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        crate::connection_cache::{get_connection, CONNECTION_MAP, MAX_CONNECTIONS},
+        fakedata_generator::gen_ipv4,
```
Comment (sakridge) [inline client/src/connection_cache.rs:92]: could we just use gen_rand on a couple of the octets? would prefer not to bring in dependencies for such a simple thing

**Review by sakridge — APPROVED** [review]
lgtm. I think we can do LruCache and removing the fake data in a follow-up

File: client/src/connection_cache.rs (hunk 4 above)
Comment (ryleung-solana) [reply client/src/connection_cache.rs]: Replaced the fakedata_generator with just the use of an RNG.

File: client/src/connection_cache.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+    let (conn, target_ticks) = match map.map.entry(*addr) {
+        Entry::Occupied(mut entry) => {
+            let mut pair = entry.get_mut();
+            let old_ticks = pair.1;
+            pair.1 = ticks;
+            (pair.0.clone(), old_ticks)
+        }
+        Entry::Vacant(entry) => {
+            let send_socket = UdpSocket::bind("0.0.0.0:0").unwrap();
```
Comment (lijunwangs) [inline client/src/connection_cache.rs:57]: Can you bind this multiple times concurrently for different addresses?

Comment (lijunwangs) [inline client/src/connection_cache.rs:57]: Can you bind this multiple times concurrently for different addresses?

Comment (t-nelson) [reply client/src/connection_cache.rs:57]: It's the same as binding to `INADDR_ANY`

Comment (lijunwangs) [reply client/src/connection_cache.rs:57]: okay -- earlier this was only bound once -- and the same send socket is used for all different addresses. Now we are create that for each unique address. That is a net change. Not sure if we need to worry about it.

**Pull request merged**